envoy/http/filter.h

```diff
@@ -802,6 +802,12 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    */
   virtual Network::Socket::OptionsSharedPtr getUpstreamSocketOptions() const PURE;
 
+  /**
+   * @param the option to set on the socket.
+   * @return true, if the option was set successfully.
+   */
+  virtual bool setDownstreamSocketOption(const Network::Socket::OptionConstSharedPtr option) PURE;
```

Comment:
Looking at this PR more closely, I think you're going through a lot of interfaces that you don't need to. For a caller that already has a `StreamDecoderFilterCallbacks`, I think you can remove all the changes in this PR except for adding a function on `Connection`, and instead of adding this function to the interface, the caller can do `stream_decoder_callbacks_->connection().value().setSocketOption(option);`

Reply:
I refactored this CL per this suggestion.